src/compiler/rust.rs

```diff
@@ -296,22 +349,56 @@ impl Rust {
             libs.sort();
             Ok((sysroot, libs))
         });
-        Box::new(sysroot_and_libs.and_then(move |(sysroot, libs)| {
+
+        #[cfg(feature = "dist-client")]
+        let rlib_dep_reader = {
+            let executable = executable.clone();
+            let env_vars = env_vars.to_owned();
+            pool.spawn_fn(move || Ok(RlibDepReader::new_with_check(executable, &env_vars)))
+        };
+
+        #[cfg(feature = "dist-client")]
+        return Box::new(sysroot_and_libs.join(rlib_dep_reader).and_then(move |((sysroot, libs), rlib_dep_reader)| {
+            let rlib_dep_reader = match rlib_dep_reader {
+                Ok(r) => Some(Arc::new(r)),
+                Err(e) => {
+                    warn!("Failed to initialise RlibDepDecoder, distributed compiles will be inefficient: {}", e);
+                    None
+                },
+            };
             hash_all(&libs, &pool).map(move |digests| {
                 Rust {
                     executable: executable,
+                    host,
                     sysroot,
                     compiler_shlibs_digests: digests,
+                    rlib_dep_reader,
                 }
             })
-        }))
+        }));
+
+        #[cfg(not(feature = "dist-client"))]
```

Comment:
I wonder if this would be more readable if you created a do-nothing future like `f_ok(None)` for the `#[cfg(not(feature = "dist-client"))]` branch, and then joined into the same code from either version?